File: recipes/subms-ts-gzip/java/src/main/java/com/submillisecond/recipes/tsgzip/Crc32.java
package com.submillisecond.recipes.tsgzip;

/**
 * CRC-32 (IEEE 802.3, polynomial 0xEDB88320), the checksum gzip stores in its
 * trailer. Hand-rolled rather than {@code java.util.zip.CRC32} so the gzip
 * layer is fully self-contained and mirrors the Rust port's
 * {@code crc32} module exactly.
 */
final class Crc32 {

    private static final int POLY = 0xEDB88320;
    private static final int[] TABLE = new int[256];

    static {
        for (int n = 0; n < 256; n++) {
            int c = n;
            for (int k = 0; k < 8; k++) {
                c = (c & 1) != 0 ? POLY ^ (c >>> 1) : c >>> 1;
            }
            TABLE[n] = c;
        }
    }

    private Crc32() {
    }

    /** CRC-32 of {@code data}, returned in the low 32 bits of a long. */
    static long crc32(byte[] data) {
        int crc = 0xFFFFFFFF;
        for (byte b : data) {
            crc = TABLE[(crc ^ b) & 0xFF] ^ (crc >>> 8);
        }
        return (crc ^ 0xFFFFFFFF) & 0xFFFFFFFFL;
    }
}

File: recipes/subms-ts-gzip/rust/src/crc32.rs
//! CRC-32 (IEEE 802.3, polynomial 0xEDB88320), the checksum gzip stores in its
//! trailer. Standard reflected-table implementation: the table is built once
//! the first time it is needed and shared via a `OnceLock`.

use std::sync::OnceLock;

const POLY: u32 = 0xEDB8_8320;

fn table() -> &'static [u32; 256] {
    static TABLE: OnceLock<[u32; 256]> = OnceLock::new();
    TABLE.get_or_init(|| {
        let mut t = [0u32; 256];
        let mut n = 0usize;
        while n < 256 {
            let mut c = n as u32;
            let mut k = 0;
            while k < 8 {
                c = if c & 1 != 0 { POLY ^ (c >> 1) } else { c >> 1 };
                k += 1;
            }
            t[n] = c;
            n += 1;
        }
        t
    })
}

/// CRC-32 of `data` over the whole buffer in one shot.
pub fn crc32(data: &[u8]) -> u32 {
    let t = table();
    let mut crc = 0xFFFF_FFFFu32;
    for &b in data {
        crc = t[((crc ^ b as u32) & 0xFF) as usize] ^ (crc >> 8);
    }
    crc ^ 0xFFFF_FFFF
}
